src/path.rs

```diff
@@ -0,0 +1,140 @@
+                    let value =
+                        value
+                            .get(idx)
+                            .ok_or_else(|| Error::Render("index out of range".to_string()))?;
+                    Ok(value)
+                }
+                (&Value::Array(_), _) => Error::renderer("bad array index type"),
+                (&Value::Object(ref value), &Value::Str(ref x)) => {
+                    let value =
+                        value
+                            .get(x)
+                            .ok_or_else(|| Error::Render("object element not found".to_string()))?;
+                    Ok(value)
+                }
+                (&Value::Object(_), _) => Error::renderer("bad object index"),
+                (value, _) if counter == 0 => Ok(value),
+                _ => Error::renderer("non-indexable element found while indexable expected"),
+            }
+        });
+
+        match result {
+            Ok(result) => result.render(context),
+            Err(e) => Error::renderer(&format!("rendering error: {}", e)),
+        }
+    }
+}
+
+impl IdentifierPath {
+    pub fn new(value: String) -> Self {
+        Self {
+            value: value,
+            indexes: Vec::new(),
+        }
+    }
+    pub fn append_indexes(&mut self, tokens: &[Token]) -> Result<()> {
+        let rest = match tokens[0] {
+            Dot if tokens.len() > 1 => {
+                match tokens[1] {
+                    Identifier(ref x) => self.indexes.push(Value::Str(x.clone())),
+                    _ => {
+                        return Error::parser("identifier", Some(&tokens[0]));
+                    }
+                };
+                2
+            }
+            OpenSquare if tokens.len() > 2 => {
+                let index = match tokens[1] {
+                    StringLiteral(ref x) => Value::Str(x.clone()),
+                    NumberLiteral(ref x) => Value::Num(*x),
+                    _ => {
+                        return Error::parser("number | string", Some(&tokens[0]));
+                    }
+                };
+                self.indexes.push(index);
+
+                if tokens[2] != CloseSquare {
+                    return Error::parser("]", Some(&tokens[1]));
+                }
+                3
+            }
+            _ => return Ok(()),
+        };
+
+        if tokens.len() > rest {
+            self.append_indexes(&tokens[rest..])
+        } else {
+            Ok(())
+        }
+    }
+}
+#[cfg(test)]
+mod test {
+    use value::Value;
+    use parse;
+    use Renderable;
+    use Context;
+    use LiquidOptions;
+    use std::collections::HashMap;
+
+    #[test]
+    fn identifier_path() {
+        let options = LiquidOptions::with_known_blocks();
+        let template = concat!("array: {{ test_a[0] }}\n",
+                               "complex_dot: {{ test_a[0].test_h }}\n",
+                               "complex_string: {{ test_a[0][\"test_h\"] }}\n");
```

Comment:
Could you make these separate tests?

Am I reading this right that the cases can be summarized as:
- identifier_path_array_index
- identifier_path_object_dot
- identifier_path_object_index

Reply:
Fixed